### PR TITLE
Created an EventMutator for when you want to mutate an event before reading

### PR DESCRIPTION
**Pull request by BobG1983**
# Objective

- Often in games you will want to create chains of systems that modify some event. For example, a chain of damage systems that handle a DamageEvent and modify the underlying value before the health system finally consumes the event. Right now this requires either:

* Using a component added to the entity
* Consuming and refiring events

Neither is ideal when really all we want to do is read the events value, modify it, and write it back.

## Solution

- Create an EventMutator class similar to EventReader but with ResMut<T> and iterators that return &mut so that events can be mutated.

## Testing

- I replicated all the existing tests for EventReader to make sure behavior was the same (I believe) and added a number of tests specific to testing that 1) events can actually be mutated, and that 2) EventReader sees changes from EventMutator for events it hasn't already seen.

## Migration Guide

Users currently using `ManualEventReader` should use `EventCursor` instead. `ManualEventReader` will be removed in Bevy 0.16. Additionally, `Events::get_reader` has been replaced by `Events::get_cursor`.

Users currently directly accessing the `Events` resource for mutation should move to `EventMutator` if possible.

### CODE REVIEW

**Comment by BobG1983** [comment]
I'm getting the same CI error here too:

2024-06-11T22:50:55.190743Z ERROR bevy_log: Could not set global logger and tracing subscriber as they are already set. Consider disabling LogPlugin.
2024-06-11T22:50:55.191809Z ERROR bevy_log: Could not set global logger and tracing subscriber as they are already set. Consider disabling LogPlugin.
2024-06-11T22:50:55.191838Z ERROR bevy_log: Could not set global logger and tracing subscriber as they are already set. Consider disabling LogPlugin.
test tests::keep_gotten_strong_handles ... ok
2024-06-11T22:50:55.199175Z ERROR bevy_log: Could not set global logger and tracing subscriber as they are already set. Consider disabling LogPlugin.
2024-06-11T22:50:55.201492Z ERROR bevy_asset::server: Encountered an I/O error while loading asset: Simulated failure 1 of 2
2024-06-11T22:50:55.204467Z ERROR bevy_asset::server: Failed to load asset 'd.cool.ron' with asset loader 'bevy_asset::tests::CoolTextLoader': A RON error occurred during loading
2024-06-11T22:50:55.208497Z ERROR bevy_log: Could not set global logger and tracing subscriber as they are already set. Consider disabling LogPlugin.
2024-06-11T22:50:55.217346Z ERROR bevy_asset::server: Encountered an I/O error while loading asset: Simulated failure 2 of 2

I haven't touched anything related to this so I'm at something of a loss.

**Comment by BobG1983** [comment]
Worked out the issue, par_read doesn't mark events as seen and thus lets you see events twice. My tests assumed par_read should behave the same as read.

For now I'll change the test.

I'll submit another PR with fixed par_read later.

**Comment by BobG1983** [comment]
To be clear, it looks like this has always been the behavior of par_read

**Comment by BobG1983** [comment]
Added the fix to #13821 but only to the par_read for EventMutator and NOT for EventReader figuring either the EventReader (PR #13836) or Added peeking to EventReader (PR #13809) will be taken which already contain the fix.

**Comment by nosideeffects** [comment]
I am not completely convinced that this addition is a strictly positive improvement in the Events API. The current expectation of most event systems is that each consumer or reader gets an ordered or unordered stream of each event since it started consuming.

Adding the ability for any number of systems to modify an event in the stream for all subsequent readers adds that property that the behavior of the event system is now order-dependent for all systems. I think it conflicts with developer intuition about how a typical event system behaves.

Consuming and re-emitting a separate event I would argue actually _is_ the ideal for adhering to the semantics of the current event system. If the need for these types of event changes proves to be fairly common, I think we would be better off providing a wrapper type that captures the entire intended behavior. For example:

```
trait EventMutator<From: Event, To: Event> {
  fn mutate_events(&mut self, mutator: FnMut(&From) -> To);
}
```
where an implementation might contain both the reader for `From` and writer for `To`. It would also make the system parameters more explicit about what mutation is actually occurring.

**Comment by BobG1983** [comment]
The problem with that is the combinatorial complexity for that is awful. DamageEvent becomes DamageAfterArmorEvent becomes DamageAfterArmorWithPoisonEvent etc. so instead I just fire off damage events and chain the systems manually anyway.

**Comment by alice-i-cecile** [comment]
I think this should exist: we have an existing example for this because users kept asking how to do this. If we can meaningfully simplify and robustify a common pattern we should here.

**Comment by nosideeffects** [comment]
I would love to see an example of this making your proposed damage system "simpler" while not making it more error-prone. In my opinion, the use of events in this way is making the system more complicated than it needs to be. The combinatorial complexity doesn't go away when you make it implicit, I think it just turns into a foot gun.

**Comment by BobG1983** [comment]
I disagree. 

This a very very common pattern in games, and systems themselves are already order dependent. If I have two systems that mutate the same component I must specify there run order or behavior can “footgun”. So this is already true. 

Not only that but I can actually already do this in a roundabout way by directly accessing the events resource this just provides better semantics and less bugs by allowing better tracking of already seeen events while unifying the API.

**Comment by nosideeffects** [comment]
Then why insist on using events at all instead of a component?

**Comment by BobG1983** [comment]
It’s more performant, it doesn’t involve adding removing components every frame, and because philosophically it *is* an event and not a component.

**Comment by BobG1983** [comment]
Again, every game does something like this. 

People have already asked how to do it often enough there is an example.

You can already do it if you know how EventReader and EventWriter are implemented. 

Why wouldn’t make it less “arcane” and provide a unified API that matches existing APIs?

**Comment by nosideeffects** [comment]
If the firing of the first event is an event, then so too are the modifications to whatever value is contained in the event. I think this muddies what the definition of an event is.

For context, I do come from a distributed systems background and not a game development background. I would prefer to see the actual morphology of the event expressible in the API. I don't think providing this in its current state is a good long-term design. If feels like it is circumventing the strengths of Rusts type system in order to keep a status quo.

**Comment by BobG1983** [comment]
I’m sorry but I don’t understand the concern. 

How is this any different from allowing component mutation?

I’m happy to listen I just don’t get it.

**Comment by nosideeffects** [comment]
Essentially, I want the API of Bevy to move in a direction where more system constraints are verifiable at compile time. Not being able to express order-or-execution constraints with component modification is not something I would say is _ideal_, it is just hard to solve. I don't really want to see even more of the API that Bevy provides out of the box to be only verifiable at runtime.

I think this change makes the behavior of the event system more ambiguous instead of less ambiguous.

**Comment by nosideeffects** [comment]
Let's say I update a plugin I use, and it now modifies the value of some emitted event in a shared dependency (or core event) before my system is able to read it. Before, perhaps my systems were order independent from the plugin. Now, however, the entire system has a different pattern of behavior even though the plugins API surface did not change.

**Comment by nosideeffects** [comment]
Would I need to implement my own event system in order to express immutable events?

**Comment by BobG1983** [comment]
Again. How is that different from component mutation?

**Comment by BobG1983** [comment]
So this a philosophical concern related to compiler vs runtime errors and mutability?

**Comment by BobG1983** [comment]
You don’t have immutable events today. What you describe can happen today.

**Comment by alice-i-cecile** [comment]
> Would I need to implement my own event system in order to express immutable events?

Yes, or add a new event kind to Bevy (like observers were added). The existing events are mutable: this is by design, and wearing both my Maintainer and SME-ECS hats, those semantics will not change. This pattern is common within game dev as a whole, and while I think that immutable events are a really interesting avenue to explore, deeply breaking existing users to support them is not on the table.

**Comment by nosideeffects** [comment]
> I think that both the existing model and immutable-by-design events can and should exist in some form, but this isn't the thread to discuss that.

I am not fundamentally against mutability or suggesting it be removed, but I also don't think making it even easier for important data flow constraints to be hidden is the right direction. I think that event mutation without an explicit API of that mutation is an anti-pattern. I guess the crux of my concern is that we are making the behavior of this particular system less explicit.

**Comment by BobG1983** [comment]
I fundamentally disagree. 

I think this is significantly more explicit than the current state of affairs.

**Comment by alice-i-cecile** [comment]
@BobG1983 can you update (I would expect nearly a full rewrite) https://github.com/bevyengine/bevy/blob/main/examples/ecs/send_and_receive_events.rs to demonstrate the net effect of this change (and also teach this feature)?

**Comment by BobG1983** [comment]
Happy to. I’ll take a look either tomorrow or Tuesday.

**Comment by BobG1983** [comment]
@alice-i-cecile I've updated the events example, but currently the send_receive_events example is still correct and the best way to do it.

I was considering adding a get_writer and get_reader to EventMutator but then I'd also expect a get_reader and get_mutator to be added to EventWriter for completion. That would require a ManualEventWriter to be created (cos EventWriter takes ownership of the resource and ResMut can't be cloned). That would then require a get_events to be added to both Mutator and Writer (and possibly Reader) and it would end up looking very similar to the existing example (with maybe a bit of a cleaner API). I think it's worth doing, but on examination, it feels like a much larger events system change and would change the scope of this PR pretty significantly.

It's my opinion that such a change would be better in it's own PR after this one is merged.

Thoughts?

**Comment by atornity** [comment]
> we have an existing example for this

Omg I had no idea, where can I find this?

File: examples/ecs/event.rs
Comment (atornity) [inline examples/ecs/event.rs:64]: what does this mean? It sounds like events must be zero sized to be sent

File: crates/bevy_ecs/src/event/mut_iterators.rs
Comment (atornity) [inline crates/bevy_ecs/src/event/mut_iterators.rs:67]: was there a reason to remove the `debug_assert_eq!(unread_count, reader.len(events)`.(from the readonly version)?

File: crates/bevy_ecs/src/event/mut_iterators.rs
Comment (atornity) [inline crates/bevy_ecs/src/event/mut_iterators.rs:94]:
```suggestion
                detailed_trace!("EventMutator::iter() -> {}", item.1);
```

File: crates/bevy_ecs/src/event/mut_iterators.rs
Comment (atornity) [inline crates/bevy_ecs/src/event/mut_iterators.rs:163]: the `debug_assert_eq!(unread_count, reader.len(events))` was removed here as well. was this intentional?

**Review by atornity — CHANGES_REQUESTED** [review]
Looks good (haven't looked at the tests). Suggested some small documentation changes.

**Comment by alice-i-cecile** [comment]
Ah @atornity, apparently you can't. Instead you have to drain, filter and then resend them all using `ResMut<Events<T>>`. Bah!

**Comment by atornity** [comment]
> Ah @atornity, apparently you can't. Instead you have to drain, filter and then resend them all using `ResMut<Events<T>>`. 

oof - I really need this functionality (specifically for a damage system)

File: crates/bevy_ecs/src/event/mut_iterators.rs
Comment (BobG1983) [reply crates/bevy_ecs/src/event/mut_iterators.rs:67]: Yeah, its provably true in all cases that I could think of and length is covered in the existing tests.

File: crates/bevy_ecs/src/event/mutator.rs
Comment (BobG1983) [reply crates/bevy_ecs/src/event/mutator.rs:184]: Will take a look. Honestly though, they're both usable in the same way here.

File: crates/bevy_ecs/src/event/mutator.rs
Comment (BobG1983) [reply crates/bevy_ecs/src/event/mutator.rs:187]: It could, would that be an issue? Would it make it less clear that you can use it with Res even if it has a read_mut function?

File: examples/ecs/event.rs
Comment (BobG1983) [reply examples/ecs/event.rs:64]: typo it should say "Even if" but actually it doesn't add anything, I'll remove it.

**Comment by BobG1983** [comment]
Give me a minute @atornity I have two more things I want to change in Events that should allow you to do this, but I have three PRs right now that all edit the same files and will need some reconciliation, so don't wanna get too ahead of myself.

File: crates/bevy_ecs/src/event/mutator.rs
Comment (atornity) [reply crates/bevy_ecs/src/event/mutator.rs:187]: I don't see any issue with that personally. 

upon reflection, I think I would simply add `read_mut` to `ManualEventReader` and use that in `EventMutator`. Or would it be weird to be able to mutate events from a type called `Reader`?

**Review by atornity — APPROVED** [review]

File: crates/bevy_ecs/src/event/mutator.rs
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/event/mutator.rs:187]: I think we should merge these types and rename it to something like EventCursor

**Comment by BobG1983** [comment]
[Created EventCursor, replacing ManualEventReader, marked ManualEventR…](https://github.com/bevyengine/bevy/pull/13818/commits/de0f0a881b562c56f57dba342cb03cc8f56711e2)
[de0f0a8](https://github.com/bevyengine/bevy/pull/13818/commits/de0f0a881b562c56f57dba342cb03cc8f56711e2)

…eader as deprecated. Updated all uses of ManualEventReader to EventCursor. Deprecated Events::get_reader and replaced it with Events::get_cursor everywhere. Also took the opportunity to tidy up the events tests. Removed unrequried tests of Mutator and Reader and replaced them with tests of EventCursor, also removed unrequried E event and replaced with TestEvent in all tests. Sorted tests so they're easier to follow

@alice-i-cecile @atornity

**Comment by BobG1983** [comment]
As an aside this now also fixes the par_read bug mentioned in #13836

**Review by atornity — APPROVED** [review]
nice, LGTM

**Comment by BobG1983** [comment]
Still waiting on a few reviews.

File: crates/bevy_ecs/src/event/event_cursor.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/event/event_cursor.rs:9]:
```suggestion
// Deprecated in favor of [`EventCursor`], there is no nice way to deprecate this
```

File: crates/bevy_ecs/src/event/mutator.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/event/mutator.rs:32]:
```suggestion
/// They also can not be executed in parallel with [`EventReader`] or [`EventWriter`].
```

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
@BobG1983 now that the release is out the door I finally have time to review this :) Very pleased with the design and engineering. Once merge conflicts are resolved I'll merge this in.

**Comment by BobG1983** [comment]
I'll get those fixed either this evening or tomorrow. 

For the migration guide. Do I add it as a comment?

**Comment by alice-i-cecile** [comment]
As part of the PR description :) I've edited the description to get it started; take a look and revise it as you see fit.

**Comment by BobG1983** [comment]
Done

**Comment by BD103** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1584 if you'd like to help out.

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1655 if you'd like to help out.

**Pull request merged**
